docker/artemis/Dockerfile

```diff
@@ -24,18 +24,44 @@ ARG WAR_FILE_STAGE="builder"
 #-----------------------------------------------------------------------------------------------------------------------
 FROM --platform=$BUILDPLATFORM docker.io/library/eclipse-temurin:17-jdk as builder
 
-WORKDIR /opt/artemis
-
 # some Apple M1 (arm64) builds need python3 and build-essential(make+gcc) for node-gyp to not fail
 RUN echo "Installing build dependencies" \
   && apt-get update && apt-get install -y --no-install-recommends python3 build-essential \
   && apt-get clean \
   && rm -rf /var/lib/apt/lists/*
 
+WORKDIR /opt/artemis
+# copy gradle related files
+COPY gradlew gradlew.bat ./
+COPY build.gradle gradle.properties settings.gradle ./
+COPY gradle gradle/
+# copy npm related files and install node modules
+# (from https://stackoverflow.com/questions/63961934/how-to-use-docker-build-cache-when-version-bumping-a-react-app)
+COPY package.json package-lock.json ./
+RUN \
+  # Mount global cache for Gradle (project cache in /opt/artemis/.gradle doesn't seem to be populated)
+  --mount=type=cache,target=/root/.gradle/caches \
+  # Mount cache for npm
+  --mount=type=cache,target=/opt/artemis/.npm \
+  # Create .npm directory if not yet available
+  mkdir -p /opt/artemis/.npm \
+  # Set .npm directory as npm cache
+  && ./gradlew -i --stacktrace --no-daemon -Pprod -Pwar npmSetCacheDockerfile \
+  # Pre-populate the npm and gradle caches if related files change (see COPY statements above)
+  && ./gradlew -i --stacktrace --no-daemon -Pprod -Pwar npm_ci
+
 # so far just using the .dockerignore to define what isn't necessary here
 COPY . .
```

Comment:
We could also mount the source folder here, so docker does not have to copy everything. Probably a follow-up.

Reply:
You mean using a bind mount instead of the copy steps in general or how would you use mounts? How do you manage the state of these mounts?